README.md
# README


**Simon Fraser University**

**CMPT 353 D100: Computational Data Science**

**Instructor: Dr. Greg Baker**

**Project Topic: OSM, Photos, and Tours**

**Group Doge**


## Background

1. This group, named "Doge", consists of Tian Xiao (txa11@sfu.ca), Yuqing Wu (ywa292@sfu.ca) and Zeyong Jin (zeyongj@sfu.ca).

2. The logo of this group is <img src="https://www.heredg.com/wp-content/uploads/2020/06/%C5%92%C2%A2%E2%80%93%E2%89%88%C3%95%C2%BA%E2%88%86%C2%A8_20200530192407.jpg" height="30" width="30">.

3. The topic of this group project is [OSM, Photos, and Tours](https://coursys.sfu.ca/2021su-cmpt-353-d1/pages/ProjectTour), check more information by clicking on the hyperlink.


## Data

1. The data is provided by the instructor, or collected online from WikiPedia and Statistics Canada.

## Libraries

This project uses the following libraries, ensure you have installed all the libraries to execute.

1. webbrowser

2. numpy

3. pandas

4. math

5. exifread

6. folium

7. sys

8. matplotlib

9. seaborn
 
10. warnings

11. os

12. urllib

13. bs4


## Instruction

1. We finished Questions 2: 
> I feel like there are some parts of the city with more chain restaurants (e.g. McDonald's or White Spot franchises, not independently-owned places): is that true? Is there some way to find the chain places automatically and visualize their density relative to non-chains?

Question 2 is done via the program of plot_density.py. You can also view the output directly via the Jupyer Notebook file of plot_density.ipynb. Further analysis is on the report.

2. We also completed Question 3: 
> If I was going to choose a hotel (or AirBnb), where should it be? What places have good amenities nearby? 

Question 3 is done via the program of airbnb_hotel.py. Further analysis is on the report.

3. Last, as of Question 4: 
> Any of these could be turned into big data problems by working with the global version of the data set. You can recreate that data as described above. 

We put forward the question about which cuisines are most welcomed in Vancouver, and based on the outcome we want to figure out whether nationality is a key factor that determines the most favourite cuisines in Vancouver. The analysis is on the report, supported by the program of plot_density.py. You can also view the output directly via the Jupyer Notebook file of plot_density.ipynb. 

## Structure

.idea

    inspectionProfiles

        profiles_settings.xml

    .gitignore

    misc.xml

    modules.xml

    osm-project.iml

    vcs.xml

    workspace.xml

code

    disassemble-osm.py

    just-vancouver.py

    osm-amenities.py

data

    .gitkeep
    
    Data_of_minority_1.csv

    amenities-vancouver.json

    amenities-vancouver.json.gz

    list_of_chain_restaurants.csv

    listings.csv


photo

    IMG_0029.JPG

    IMG_0030.JPG

    IMG_0031.JPG

    IMG_0033.JPG

    IMG_0037.JPG

    IMG_0039.JPG

    IMG_0044.JPG

    IMG_2628.JPG

    IMG_2641.JPG

    IMG_5861.JPG


results

    f1_density.jpg
    
    f2_chain.jpg

    f3_non-chain.jpg

    f4_amenities.jpg

    f5_top20.jpg

    f6_cuisine.jpg

    f7_piecuisine.jpg

    f8_citizenship.jpg

    f9_piecitizenship.jpg

    hotel.csv

    hotel_map.html

CHANGELOG.md

CONTRIBUTING.md

LICENSE

Project_Report_Group_Doge.pdf

README.md

airbnb_hotel.py

hotel_map.html

plot_density.ipynb

plot_density.py


## Execution

1. To run plot_density.py, type the following in the terminal: `python plot_density.py`.

    1.1 The program will generate a csv file of list_of_chain_restaurant.csv. You can directly view output from the folder of data, or via the Jupyter Notebook file of plot_density.ipynb. 

    1.2 The program will generate 9 JPG pictures, named from f1 to f9. You can directly view output from the folder of results, or via the Jupyter Notebook file of plot_density.ipynb. 
   
    1.3 The pictures would be saved on the folder of results.

2. To run airbnb_hotel.py, type the following in the terminal: `python airbnb_hotel.py photo/IMG_0029.JPG all`. 

   2.1 You can change  `IMG_0029.JPG` to any of JPG files in the photo folder.

   2.2 `all` can be changed to any integer, since `price` is integer in the "listings.csv" file.  
   
   2.3 The program will generate a html file of hotel_map.html as well as a csv file of hotel.csv. You can directly view the hotel_map.html via a web browser from the folder of results. After execution, the system would also open the hotel_map.html automatically for you.

## Contribution

1. Yuqing Wu (ywa292@sfu.ca) and Zeyong Jin (zeyongj@sfu.ca) were responsible for Questions 2 and 4.

2. Tian Xiao (txa11@sfu.ca) finished the part of Question 3.

3. Tian Xiao (txa11@sfu.ca), Yuqing Wu (ywa292@sfu.ca) and Zeyong Jin (zeyongj@sfu.ca) wrote the report by cooperation.

## Versions

1. This project was created on July 24th, 2021.

2. This project was last modified and submitted to the CSIL GitLab platform on August 13th, 2021.


## License

1. This work is licensed under [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0) (or any later version).

    `SPDX-License-Identifier: Apache-2.0-or-later`

## Disclaimer

**This repository is *ONLY* for backup. Students should *NEVER* use this repository to finish their works, *IN ANY WAY*.**

It is expected that within this course, the highest standards of academic integrity will be maintained, in
keeping with SFU’s Policy S10.01, `Code of Academic Integrity and Good Conduct`.

In this class, collaboration is encouraged for in-class exercises and the team components of the assignments, as well
as task preparation for group discussions. However, individual work should be completed by the person
who submits it. Any work that is independent work of the submitter should be clearly cited to make its
source clear. All referenced work in reports and presentations must be appropriately cited, to include
websites, as well as figures and graphs in presentations. If there are any questions whatsoever, feel free
to contact the course instructor about any possible grey areas.

Some examples of unacceptable behaviour:
- Handing in assignments/exercises that are not 100% your own work (in design, implementation,
wording, etc.), without a clear/visible citation of the source.
- Using another student's work as a template or reference for completing your own work.
- Using any unpermitted resources during an exam.
- Looking at, or attempting to look at, another student's answer during an exam.
- Submitting work that has been submitted before, for any course at any institution.

All instances of academic dishonesty will be dealt with severely and according to SFU policy. This means
that Student Services will be notified, and they will record the dishonesty in the student's file. Students
are strongly encouraged to review SFU’s Code of Academic Integrity and Good Conduct (S10.01) available
online at: http://www.sfu.ca/policies/gazette/student/s10-01.html.


## Epilogue

1. Group Doge presents its compliments to the instructor Dr. Baker and teaching assistants in the course of computational data science.

2. If you have any questions or suggestions, please contact us via the aforementioned emails.



## Authors

Group Doge  <img src="https://www.heredg.com/wp-content/uploads/2020/06/%C5%92%C2%A2%E2%80%93%E2%89%88%C3%95%C2%BA%E2%88%86%C2%A8_20200530192407.jpg" height="30" width="30">

August 25th, 2021


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [OSM, Photos, and Tours](https://coursys.sfu.ca/2021su-cmpt-353-d1/pages/ProjectTour).

## [5.0.2] - 2021-08-13
 
Update files.

### Added


### Changed

- results

- CHANGELOG.md

### Fixed



## [5.0.1] - 2021-08-13
 
Fix bugs on the airbnb_hotel.py.

### Added


### Changed

- CHANGELOG.md

### Fixed

- airbnb_hotel.py


## [5.0.0] - 2021-08-13
 
Finished the project report, modifed README.md.

### Added

- Project_Report_Group_Doge.pdf

### Changed

- README.md

- CHANGELOG.md

### Fixed



## [4.0.1] - 2021-08-11
 
Combine codes, fix bugs.

### Added

- hotel_map.html

### Changed

- plot_density.py

- plot_density.ipynb

- photo

- data

- results

- README.md

- CHANGELOG.md

### Deleted

- get_restaurant_wiki.py

- get_restaurant_wiki.ipynb

### Fixed

- plot_density.py

- airbnb_hotel.py

## [4.0.0] - 2021-08-11
 
Reset the directories, rearrnage files, fix bugs.

### Added

- hotel_map.html

### Changed

- results

- photo

- data

### Fixed

- plot_density.py

- airbnb_hotel.py


## [3.0.3] - 2021-08-11
 
Here we added the result of Question 3.

### Added

- hotel_map.html

### Changed


### Fixed



## [3.0.2] - 2021-08-11

Here we added some photos for airbnb_hotel.py and fix some bugs.

### Added

- IMG_0029.JPG

- IMG_0030.JPG

- IMG_0031.JPG

- IMG_0033.JPG

- IMG_0037.JPG

- IMG_0039.JPG

- IMG_0044.JPG

- IMG_2628.JPG

- IMG_2641.JPG

- IMG_5861.JPG

### Changed

- README.md

- CHANGELOG.md

### Fixed

- airbnb_hotel.py

## [3.0.1] - 2021-08-11
 
Here we finished airbnb_hotel.py for Question 3 and fix some bugs.

### Added


### Changed

- README.md

- airbnb_hotel.py

- CHANGELOG.md

### Fixed

- airbnb_hotel.py

 
## [3.0.0] - 2021-08-11
 
Here we write airbnb_hotel.py for Question 3.

### Added

- airbnb_hotel.py

### Changed

- README.md

- CHANGELOG.md

### Fixed



## [2.0.3] - 2021-08-09
  
Here we provided Jupyter Notebook's ipynb files for audience to view the outcome easily.
 
### Added

- get_restaurant_wiki.ipynb

- plot_density.ipynb
 
### Changed


### Fixed


## [2.0.2] - 2021-08-09
  
Here we wrote some files and debugged.
 
### Added

- LICENSE

- CONTRIBUTING.md

- CHANGELOG.md
 
### Changed

- Modifed the README file.
 
### Fixed
- Fix the bug of figure display.



## [2.0.1] - 2021-08-09
  
Here we solved Question 4.
 
### Added

- Datasets
 
### Changed

- Modifed the plot_density, delete the population_density but added a feature of analyzing nationality.
 
### Fixed
- Fix the bug of figure display.
 

## [2.0.0] - 2021-08-08
  
Here we basically solved Question 2.
 
### Added

- Datasets
 
### Changed
  
- Modified the plot_density to create figures for Question 4.
 
### Fixed
 
 
## [1.0.1] - 2021-08-04
  
Here we solved Question 2.
 
### Added

 
### Changed
  
 
### Fixed
- Fix the bug of figure display.
 
## [1.0.0] - 2021-08-03
  
Here we basically solved Question 2.
 
### Added

- population_density.py
- Datasets
 
### Changed
  
- Modified the plot_density to create figures for Question 2.
 
### Fixed
 
 
## [0.0.1] - 2021-07-24

Here we started the project.
 
### Added

- Provided files
- Datasets
- plot_density.py
- README.md
   
### Changed
 
### Fixed


airbnb_hotel.py
import webbrowser
import numpy as np
import pandas as pd
from math import pi, cos
import exifread
import folium
from folium.plugins import HeatMap
import sys

# the listings.csv are collected from http://insideairbnb.com/get-the-data.html
# the code for coordinates is adapted from http://insideairbnb.com/get-the-data.html

def get_coordinates(photo):
    with open(photo, 'rb') as f:
        exif_dict = exifread.process_file(f)
        lon_ref = exif_dict["GPS GPSLongitudeRef"].printable
        lon = exif_dict["GPS GPSLongitude"].printable[1:-1].replace(" ", "").replace("/", ",").split(",")
        lon = float(lon[0]) + float(lon[1]) / 60 + float(lon[2]) / float(lon[3]) / 3600
        if lon_ref != "E":
            lon = lon * (-1)
        lat_ref = exif_dict["GPS GPSLatitudeRef"].printable
        lat = exif_dict["GPS GPSLatitude"].printable[1:-1].replace(" ", "").replace("/", ",").split(",")
        lat = float(lat[0]) + float(lat[1]) / 60 + float(lat[2]) / float(lat[3]) / 3600
        if lat_ref != "N":
            lat = lat * (-1)
    return lat, lon


def distance(hotel_range):
    p = pi / 180
    lat1 = hotel_range['lat']
    lon1 = hotel_range['lon']
    lat2 = hotel_range['latitude']
    lon2 = hotel_range['longitude']
    a = 0.5 - np.cos((lat2 - lat1) * p) / 2 + np.cos(lat1 * p) * np.cos(lat2 * p) * (1 - np.cos((lon2 - lon1) * p)) / 2
    b = 12742 * np.arcsin(np.sqrt(a)) * 1000
    return b


def has_tourism(tags):
    return 'tourism' in tags


def getdata(frame):
    temp = frame[frame['tags'].apply(has_tourism)]
    temp_of_four = frame[(frame['amenity'] == 'food_court') |
                                (frame['amenity'] == 'atm') |
                                (frame['amenity'] == 'bus_station') |
                                (frame['amenity'] == 'clinic')|
                                (frame['amenity'] == 'fast_food')]
    frames = [temp, temp_of_four]
    result = pd.concat(frames)
    return result


def main(in_directory,  max_budget):
    data = pd.read_csv("data/listings.csv")
    amenity_data = pd.read_json('data/amenities-vancouver.json.gz', lines=True)
    heat = getdata(amenity_data)
    photo = in_directory
    lat, lon = get_coordinates(photo)

    earth = 6378.137
    #Using the coordinates of one point and the distance we want to get the hotel list
    #https://stackoverflow.com/questions/7477003/calculating-new-longitude-latitude-from-old-n-meters
    m = (1 / ((2 * pi / 360) * earth)) / 1000
    new_lat1 = lat + (300 * m)
    new_lat2 = lat - (300 * m)

    new_lon1 = lon + (300 * m) / cos(lat * (pi / 180))
    new_lon2 = lon - (300 * m) / cos(lat * (pi / 180))

    hotel_range = data.loc[data['longitude'] < new_lon1]
    hotel_range = hotel_range.loc[hotel_range['longitude'] > new_lon2]
    hotel_range = hotel_range.loc[hotel_range['latitude'] < new_lat1]
    hotel_range = hotel_range.loc[hotel_range['latitude'] > new_lat2]
    if(max_budget!="all"):
        hotel_range = hotel_range.loc[hotel_range['price'] <= int(max_budget)]

    hotel_range = hotel_range.reset_index(drop=True)
    hotel_range['lon'] = lon
    hotel_range['lat'] = lat
    hotel_range['distance between'] = distance(hotel_range)
    hotel_range = hotel_range.drop(['lon', 'lat'], axis=1)
    hotel_range = hotel_range.sort_values(by='distance between')
    hotel_range.to_csv('results/hotel.csv')

    #map drawing from: https://zhuanlan.zhihu.com/p/112324234
    van_map = folium.Map(location=[lat, lon], zoom_start=16)
    lat_of_me = lat
    lon_of_me = lon
    incidents = folium.map.FeatureGroup()
    latitudes = list(hotel_range.latitude)
    longitudes = list(hotel_range.longitude)
    labels = list(hotel_range.name)
    for lat, lng, label in zip(latitudes, longitudes, labels):
        folium.Marker([lat, lng], popup=label).add_to(van_map)
    van_map.add_child(incidents)
    folium.Marker(
        location=[lat_of_me, lon_of_me],
        popup="where am I",
        icon=folium.Icon(color='red', icon="cloud" ),
    ).add_to(van_map)
    heatpoints = heat[['lat', 'lon']].values.tolist()
    HeatMap(heatpoints).add_to(van_map)
    van_map.save('results/hotel_map.html')
    van_map.save('hotel_map.html')
    webbrowser.open('hotel_map.html', new=2)
    # import os
    # os.remove("hotel_map.html")

if __name__ == '__main__':
    in_directory = sys.argv[1]
    max_budget = sys.argv[2]
    main(in_directory, max_budget)


plot_density.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from math import *
import seaborn
import warnings
import os
warnings.filterwarnings("ignore")
seaborn.set()
import urllib.parse
import urllib.request
from bs4 import BeautifulSoup
import pandas as pd

# Get restaurants from wiki first.
# the code are adapted from https://blog.csdn.net/nmjuzi/article/details/113766630
chain_restaurants = []
output = []
url_canadian_restaurant_chains = 'https://en.wikipedia.org/wiki/List_of_Canadian_restaurant_chains'
wiki_canadian_restaurant_chains = urllib.request.urlopen(url_canadian_restaurant_chains)
get_canadian_restaurant_chains_html = wiki_canadian_restaurant_chains.read()
get_canadian_restaurant_chains_html = get_canadian_restaurant_chains_html.decode('utf-8')
get_canadian_restaurant_chains_html_soup = BeautifulSoup(get_canadian_restaurant_chains_html, 'html.parser')
answers = get_canadian_restaurant_chains_html_soup.find_all('span', class_='mw-headline')
for restaurant in answers:
	text = restaurant.get_text()
	output.append(text)
output_length = len(output)
upper_bound = output_length - 3
lower_bound = 1
output = output[lower_bound:upper_bound]
for restaurant in output:
	seg = '('
	name = restaurant.split(seg)
	chain_restaurants.append(name[0])
url_restaurant_chains = 'https://en.wikipedia.org/wiki/List_of_restaurant_chains'
wiki_restaurant_chains = urllib.request.urlopen(url_restaurant_chains)
get_restaurant_chains_html = wiki_restaurant_chains.read()
get_restaurant_chains_html = get_restaurant_chains_html.decode('utf-8')
pageSoup = BeautifulSoup(get_restaurant_chains_html, 'html.parser')
answer = pageSoup.find('table', class_='wikitable sortable')
answers = answer.find_all('tr')
for restaurant in answers:
	name = restaurant.find('a',href = True,title = True)
	if name is not None:
		chain_restaurants.append(name.get_text())
chain_restaurants.sort(key = str.lower) # Adapted from https://www.learnbyexample.org/python-list-sort-method/ .
chain_restaurants = list(dict.fromkeys(chain_restaurants))
pd_chain = pd.DataFrame(chain_restaurants)
pd_chain.to_csv('./data/list_of_chain_restaurants.csv', index = False, header = False)

amenities=pd.read_json('./data/amenities-vancouver.json.gz', lines = True)
chain_restaurants=pd.read_csv('./data/list_of_chain_restaurants.csv')
chain_restaurants['list'] = chain_restaurants

# the code is adapted from https://blog.csdn.net/u014769180/article/details/83896266
def in_list(example):
	if example in list(chain_restaurants['list']):
		return True
	else:
		return False

def sep_chain_nonchain(example):
	restaurant_list=example[(example['amenity']=='bbq')|(example['amenity']=='restaurant')|(example['amenity']=='fast_food')
	|(example['amenity']=='cafe')|(example['amenity']=='bar')|(example['amenity']=='juice_bar')|(example['amenity']=='food_court')]
	in_chain=restaurant_list[restaurant_list['name'].apply(in_list)]
	out_chain=restaurant_list[~restaurant_list['name'].apply(in_list)]
	return in_chain,out_chain

in_chain, out_chain=sep_chain_nonchain(amenities)

# Figure 1
# the scatter code is adapted from https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.scatter.html
plt.xlabel('longitude')
plt.ylabel('latitude')
plt.title('Figure 1: Distribution of Restaurants in Vancouver')
plt.scatter(out_chain['lon'], out_chain['lat'], c="orange", s=10)
plt.scatter(in_chain['lon'], in_chain['lat'], c="purple", s=10)
plt.legend(['non-chain', 'chain'])
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
strFile = "./results/f1_density.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()

# Figure 2
# the hist2d code is adapted from https://www.geeksforgeeks.org/plot-2-d-histogram-in-python-using-matplotlib/
plt.xlabel('longitude')
plt.ylabel('latitude')
plt.title('Figure 2: Density of Chain Restaurants in Vancouver')
plt.hist2d(in_chain['lon'], in_chain['lat'])
plt.colorbar()
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
strFile = "./results/f2_chain.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()

# Figure 3
plt.xlabel('longitude')
plt.ylabel('latitude')
plt.title('Figure 3: Density of Non-chain Restaurants in Vancouver')
plt.hist2d(out_chain['lon'], out_chain['lat'])
plt.colorbar()
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
strFile = "./results/f3_non-chain.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()

# Figure 4
summary=amenities.groupby('amenity').count()
summary=summary.drop(['timestamp','tags','lon','name'], axis=1)
summary=summary.rename(columns={"lat":"number"}).reset_index()
plt.figure(figsize=(20,15))
plt.xlabel('amenities')
plt.ylabel('frequency')
plt.title('Figure 4: Histogram of Amenities Frequency in Vancouver')
plt.bar(summary['amenity'],summary['number'])
plt.xticks(size=6.5,rotation='vertical')
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
strFile = "./results/f4_amenities.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()

amenities['lat_lon'] = amenities[['lat', 'lon']].values.tolist()
# the code are adapted from https://www.huaweicloud.com/articles/7a48781dc78aa4aa5192394f46300838.html
def check_restaurant(lat_lng,example):
	lat=lat_lng[0]
	lng=lat_lng[1]
	r=6371
	dis=0.5
	dlng=2*asin(sin(dis/(2*r))/cos(lat*pi/180))
	dlng=dlng*180/pi
	dlat=dis/r
	dlat=dlat*180/pi
	max_lat=lat + dlat
	min_lat=lat - dlat
	max_lon=lng + dlng
	min_lon=lng - dlng
	result=example[(example['lat']>=min_lat)&(example['lat']<=max_lat)&(example['lon']>=min_lon)&(example['lon']<=max_lon)]
	return result.count()[0]

def get_num(name, example):
	amenity=amenities[amenities['amenity']==name]
	amenity['count']=amenity['lat_lon'].apply(check_restaurant, args=(example,)) # the code are adapted from https://blog.csdn.net/jewely/article/details/107888098
	return amenity[['amenity','count']]

not_restaurants=amenities[(amenities['amenity']!='bbq')&(amenities['amenity']!='restaurant')&(amenities['amenity']!='fast_food')
    &(amenities['amenity']!='cafe')&(amenities['amenity']!='bar')&(amenities['amenity']!='juice_bar')&(amenities['amenity']!='food_court')]
summary=not_restaurants.groupby('amenity',as_index=False).count()
summary=summary.drop(['timestamp','tags','lat','name','lat_lon'], axis=1)
summary=summary.rename(columns={"lon":"number"}).reset_index()
summary=summary[(summary['number']>=100)]
chains=pd.DataFrame()
non_chains=pd.DataFrame()
for amenity in list(summary['amenity']):
    c=get_num(amenity, in_chain)
    n=get_num(amenity, out_chain)
    chains=chains.append(c)
    non_chains=non_chains.append(n)

# Figure 5
chains=chains.groupby('amenity',as_index=False)['count'].mean()
non_chains=non_chains.groupby('amenity',as_index=False)['count'].mean()
plt.figure(figsize=(12,9))
# the double histgram are adapted from https://www.jb51.net/article/185341.htm
bar_width=0.3
index_chain=np.arange(len(chains['amenity']))
index_notchain=index_chain + bar_width
plt.bar(index_chain,height=list(chains['count']),width=bar_width,color='r', label='chain')
plt.bar(index_notchain,height=list(non_chains['count']),width=bar_width,color='b', label='non_chain')
plt.xticks(index_chain+bar_width/2, list(chains['amenity']))
plt.legend()
plt.xticks(size=6.5,rotation='vertical')
plt.xlabel('amenities')
plt.ylabel("Average number of restaurants")
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
plt.title("Figure 5: Top 20 Amenities of The Number of Restaurants Within 0.5 Km")
strFile = "./results/f5_top20.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()

def has_cuisine(tags):
	return 'cuisine' in tags
temp=amenities[amenities['tags'].apply(has_cuisine)]
name=list(temp['amenity'])
cuisine=list(temp['tags'])
types=pd.DataFrame(columns=['cuisine','count'])
for i in range(len(cuisine)):
	cuisine[i]=cuisine[i]['cuisine']
	title=str(cuisine[i]).lower()
	title=title.split(';')
	if len(title)==1:
		types=types.append({"cuisine": title[0],"count":int(1)},ignore_index=True)
	else:
		for j in range(len(title)):
			types=types.append({"cuisine": title[j],"count":int(1)},ignore_index=True)

# Figure 6
types=types.groupby('cuisine',as_index=False).sum()
example=types[(types['count']>=15)]
plt.figure(figsize=(12,9))
plt.xlabel('cuisine')
plt.ylabel("number")
plt.title("Figure 6: Histogram of Cuisine Kinds in Vancouver")
plt.bar(example['cuisine'],example['count'])
plt.xticks(size = 7,rotation = 45)
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
strFile = "./results/f6_cuisine.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()

# Figure 7
# the pie chart code are adapted from https://www.jb51.net/article/173450.htm
example.loc[example.cuisine =='noodle','cuisine']="chinese"
example.loc[example.cuisine =='sushi','cuisine']="japanese"
example.loc[example.cuisine =='bubble_tea','cuisine']="drink"
example.loc[example.cuisine =='kebab','cuisine']="indian"
example.loc[example.cuisine =='juice','cuisine']="drink"
example.loc[example.cuisine =='coffee_shop','cuisine']="drink"
example.loc[example.cuisine =='pizza','cuisine']="american"
example.loc[example.cuisine =='sandwich','cuisine']="american"
example.loc[example.cuisine =='burger','cuisine']="american"
example=example.groupby('cuisine',as_index=False).sum()
plt.figure(figsize=(15,9))
plt.title("Figure 7: Pie Chart of Processed Cuisine Kinds in Vancouver")
plt.pie(example['count'],labels=example['cuisine'],radius=80,autopct='%1.2f%%',textprops={'fontsize':8,'color':'k'})
plt.axis('equal')
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
strFile = "./results/f7_piecuisine.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()

# Figure 8
# the data are collected from https://www12.statcan.gc.ca/census-recensement/2016/dp-pd/dt-td/Rp-eng.cfm?TABID=2&LANG=E&A=R&APATH=3&DETAIL=0&DIM=0&FL=A&FREE=0&GC=59&GL=-1&GID=1341689&GK=1&GRP=1&O=D&PID=112048&PRID=10&PTYPE=109445&S=0&SHOWALL=0&SUB=0&Temporal=2017&THEME=120&VID=0&VNAMEE=&VNAMEF=&D1=0&D2=0&D3=0&D4=0&D5=0&D6=0#fnb2
minority=pd.read_csv("data/Data_of_minority_1.csv",encoding='unicode_escape')
minority.loc[1,'Total']=136905
minority=minority.drop([27])
top=minority.sort_values("Total",ascending=False)
top=top[0:10]
top=top[['Countries_of_citizenship','Total']]
plt.figure(figsize=(15,9))
plt.xlabel('Countries of Citizenship')
plt.ylabel("Total")
plt.title("Figure 8: Histogram of Countries of Citizenship in Vancouver")
plt.bar(top['Countries_of_citizenship'],top['Total'])
plt.xticks(size = 7, rotation = 45)
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
strFile = "./results/f8_citizenship.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()

# Figure 9
plt.figure(figsize=(15,9))
plt.pie(top['Total'],radius=80,autopct='%1.2f%%',textprops={'fontsize':6,'color':'k'})
elements=top['Countries_of_citizenship']
plt.legend(elements,
           fontsize=12,
           title="Countries of Citizenship",
           loc="center left",
           bbox_to_anchor=(0.91, 0, 0.3, 1))
plt.axis('equal')
# The following codes are adapted from https://stackoverflow.com/questions/49039581/matplotlib-savefig-will-not-overwrite-old-files .
plt.title("Figure 9: Pie Chart of Countries of Citizenship in Vancouver")
strFile = "./results/f9_piecitizenship.jpg"
if os.path.isfile(strFile):
   os.remove(strFile)   # Opt.: os.system("rm "+strFile)
plt.savefig(strFile)
plt.close()
